docs/expiry-and-auto-renewal/renewal_expiry_design_v2.md

```diff
@@ -0,0 +1,279 @@
+
+			if (!action.isEmpty() && hasAvailableCapacity(availability, action.requiredCapacity())) {
+              			action.execute();
+              			availability -= action.requiredCapacity() + action.requiredCapacity() * throttleLoad();
+              			lastScannedEntity = accountNum;
+			}
+		}
+
+		return available;
+	}
+}
+````
+
+The implementation of `AccountDeletionBucket` will be responsible for:
+
+- iteration over all account relations
+- for every iteration (entity touch) the capacity will be depleted
+- determination of action for each merkle account loaded from state
+- action execution
+- removal of account from `MerkleNetworkContext` queue
+
+Pseudo code of the drain method:
+
+````java
+class AccountDeletionBucket {
+	public Capacity drain(long consensusTimeStamp, Capacity available) {
+		while (hasEnoughCapacity(available)) {
+			EntityId accountId = getNextRelationForProcessing();
+			MerkleAccount merkleAccount = accounts.get(accountId);
+			Action action = determineAction(merkleAccount);
+
+			if (hasAvailableCapacity(availability, action.requiredCapacity())) {
+				action.execute();
+				available -= action.requiredCapacity();
+			}
+		}
+
+		return available;
+	}
+}
+````
+
+-------------
+
+### Action
+
+An action is an abstraction of an executable task (business logic) which should be performed in order to expire/renew an
+account in atomic way. Each action should be scoped to a single entity. Some actions will be too time consuming to be
+executed as part of Hedera transaction. An example for that is an account that has expired and the grace period has
+passed.
+
+For migrating the current expiry/renewal account implementation, we might consider 3 types of actions
+- `AccountScannerAction`, `AccountRemovalAction` and `TokenCleanupAction`.
+
+Below is `Action` functional interface:
+
+```java
+interface Action {
+	/* Returns the required capacity of the Action */
+	Capacity requiredCapacity();
+
+	/* Performs the action */
+	void execute();
+}
+```
+
+The implementation `AccountScannerAction` will be responsible for:
+
+- add account for reletationship removal to `MerkleNetworkContext` new `Queue<AccountNum>`
+
+Pseudo code of execute method:
+
+````java
+class AccountScannerAction {
+	public execute() {
+		merkleNetworkContext.addAccountForRelRemoval(accountNum);
+	}
+}
+````
+
+The implementation `AccountRemovalAction` will be responsible for:
+
+- removal from `FCMap<MerkleEntityId, MerkleAccount> accounts`
+- remove account from `MerkleNetworkContext` new `Queue<AccountNum>`
+
+Pseudo code of execute method:
+
+````java
+class AccountRemovalAction {
+	public execute() {
+		FCMap<MerkleEntityId, MerkleAccount> accounts = getAccounts();
+		accounts.remove(accountNum);
+		merkleNetworkContext.removeAccountForRelRemoval(accountNum);
+	}
+}
+````
+
+The implementation `TokenCleanupAction` will be responsible for:
+
+- iterate over account tokens while checking for available capacity
```

Comment:
The action must not iterate through the tokens. The action should have a single token rel to remove. This is done in the bucket.